### ISSUE TITLE
x/tools/internal/lsp: handle module paths below root path

### ISSUE THREAD

**Issue by eliasnaur** [golang/go#29174]
I'm trying to get golsp to work with vim through https://github.com/autozimu/LanguageClient-neovim.

Golsp seems to work very well for me when editing Go source code in GOPATH. However, when editing Go source inside a module, golsp reports

"no packages found for </path/to/gofile.go>"

when I attempt the "go to definition" command.

My project structure resembles this template:

```
project/
 .git/
 src/
   example.com/
     go.mod
     source.go

```

It seems that the problem is a mismatch between the language server "project path" and the go module path.

The project path is sent through the "initialize" lsp request in the "rootPath" and "rootUri" fields by LanguageClient-neovim. LanguageClient-neovim also starts the golsp executable with the project path as its current directory. Its unclear exactly how the project path is determined, but it seems to be the root of the git project containing the Go source file I'm editing. In my case, "project/".

Now, when golsp attempts to determine the packages containing for the open Go source file it fails because

`pkgs, err := packages.Load(v.Config, fmt.Sprintf("file=%s", path))`

(from https://github.com/golang/tools/blob/6a3e9aa2ab7749d72d1006ee484271b4a11f96c2/internal/lsp/cache/view.go#L63)

fails to find the current module in golsp's current working directory, "project/".

To test my hypothesis, I added

`Dir: "</path/to/project/src/example.com>"`

to the packages.Config used above and sure enough, "go to definition" and other lsp features started working again.

I'm sorry if this is a known issue. If so, I hope this issue can be a placeholder for resolving it.

CC @stamblerre.

**Comment by stamblerre** [comment]
thanks for reporting this @eliasnaur! This is definitely the correct fix; I think none of us just got around to fixing it because we are working in $GOPATH mode. If you'd like to contribute the fix, please go ahead, otherwise I'll be happy to fix it.

**Comment by myitcv** [comment]
@eliasnaur - very interesting, we were talking about editors gaining LSP capabilities on yesterday's [`golang-tools`](https://github.com/golang/go/wiki/golang-tools) call.

If it's ok with you I'll cc you on a thread I'd planned to kick off about Vim/Neovim integration with the LSP?

**Comment by eliasnaur** [comment]
Sure, go ahead.

ons. 12. dec. 2018 14.43 skrev Paul Jolly <notifications@github.com>:

**Comment by eliasnaur** [comment]
Thanks for the CL, but it doesn't actually fix the actual problem I reported: the LanguageClient-neovim vim plugin will pass my git project root as rootUri; the module is below the git project root and golsp will fail to locate package. Who's at fault, golsp or LanguageClient-neovim?

**Comment by eliasnaur** [comment]
But changing the vim plugin seems to go against the stated goal of the language server protocol: reducing an M*N problem to M + N. Case in point, LanguageClient-neovim doesn't know about golsp or any other particular language server; I mapped go files to the golsp server in my vim configuration.
Furthermore, it seems odd to me that the rootUri is established at the "initialize" request, but a given workspace (git project) might very well contain multiple modules. If so, is the lsp client supposed to re-initialize its lsp server for every file in a different module? Or have a server running for each Go module?

In other words, are you suggesting that every lsp client, including the multiple clients for vim alone, must include Go specific logic to determine rootUri?
And even if your answer is yes, I think it is worth having a fallback in golsp for a (from the point of view of golsp) "wrong" rootUri. I think a "no packages found for </path/to/gofile.go>" is giving up too easily.

**Comment by fsouza** [comment]
@eliasnaur off-topic, but you can fix this in LanguageClient-neovim using a configuration parameter (use ``go.mod`` as the root marker for Go projects). See ``g:LanguageClient_rootMarkers``.

**Comment by benweissmann** [comment]
To add another data point: I'm using VSCode and working on a project that contains multiple Go modules. `gopls` works great as long as I open one of the modules directly, but if I open the parent folder (which it not itself a Go modules, but has sub-directories that are Go modules) then `gopls` doesn't work (can't use go-to-definition, etc.). It also doesn't work if I add the modules separately as top-level folders in the workspace (using `File -> Add Folder To Workspace`) -- `gopls` works with the first folder in the workspace but not the others. If I open up a separate VSCode window for each module, it does work.

**Comment by stamblerre** [comment]
Duplicate of #30841

**Comment by stamblerre** [comment]
More discussion seems to be happening on #30841, closing in favor of that issue
